### PR TITLE
[Hub Generated] Review request for Microsoft.DBforPostgreSQL to add version stable/2022-11-08

### PR DESCRIPTION
**Pull request by senatol**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/dev-postgresqlhsc-Microsoft.DBforPostgreSQL-2022-11-08...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
**End of January 2023**
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
**February 2023**
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.
**We haven't generated SDK until now. It's the first version we are going GA with. Therefore we don't need to refresh any SDK versions.**

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by senatol** [comment]
Note to ARMTeam: This version will be the first version that will be GA. The private-preview version which was used as base for this PR is not publicly available and not used by any customer, it was mistakenly pushed to repository in the past. So the breaking changes shouldn't apply in our case. Can you help with the breaking changes tag?

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 1]
```diff
@@ -0,0 +1,2987 @@
...
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Cluster"
+            },
+            "description": "The required parameters for creating or updating a cluster."
+          }
+        ],
+        "responses": {
+          "201": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:164]: >": [](http://example.com/codeflow?start=14&length=2)

you still need to keep the 200 response which should be returned for a RE-PUT scenario

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 2]
```diff
@@ -0,0 +1,2987 @@
...
+          {
+            "$ref": "../../../../../common-types/resource-management/v4/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "#/parameters/ClusterNameParameter"
+          }
+        ],
+        "responses": {
+          "202": {
+            "description": "Accepted",
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:257]: Need to keep the 200 for the final status returned by the locaiton header

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 3]
```diff
@@ -0,0 +1,2987 @@
...
+            "schema": {
+              "$ref": "#/definitions/ClusterForUpdate"
+            },
+            "description": "The parameters for updating a cluster."
+          }
+        ],
+        "responses": {
+          "202": {
+            "description": "Accepted",
+            "headers": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:331]: >headers [](http://example.com/codeflow?start=13&length=7)

same comment about the 200 return value everywhere in this PR

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 4]
```diff
@@ -0,0 +1,2987 @@
...
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/serverGroupsv2/{clusterName}/configurations": {
+      "get": {
+        "description": "List all the configurations of a cluster.",
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:492]: >cluster [](http://example.com/codeflow?start=57&length=7)

I do not recommend making this change , the URL still is calling it a serverGroupV2 and this descrption makes it confusing for a user

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 5]
```diff
@@ -0,0 +1,2987 @@
...
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v4/types.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/serverGroupsv2/{clusterName}/configurations/{configurationName}/node": {
+      "put": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:647]: >put [](http://example.com/codeflow?start=7&length=3)

you cannot do a PUT like his on a resource-collection. 

so in ARM the URLS have the following format : 

resource-collection\resourcename\resource-collection\resourcename

the way this is defined here it will be interpreted as a PUT on a resource collection called "node" which is not allowed

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 6]
```diff
@@ -0,0 +1,2987 @@
...
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Role"
+            },
+            "description": "The required parameters for creating or updating a role."
+          }
+        ],
+        "responses": {
+          "202": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:964]: >": [](http://example.com/codeflow?start=14&length=2)

please fix the 200 issue everywhere in this PR

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 7]
```diff
@@ -0,0 +1,2987 @@
...
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v4/types.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/serverGroupsv2/{clusterName}/roles": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1042]: >clusterName [](http://example.com/codeflow?start=124&length=11)

as mentioned above please revert this change of renaming the clusterName variable to serverGroupName variable everywhere in this PR

Comment (raosuhas) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1042]: you can keep this change if you also rename the resourcetype from serverGroupsv2 to cluster or something appropriate

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 8]
```diff
@@ -0,0 +1,2987 @@
...
+        },
+        "coordinatorVCores": {
+          "type": "integer",
+          "format": "int64",
+          "description": "The vCores count of the coordinator (max: 64)."
+        },
+        "coordinatorEnablePublicIpAccess": {
+          "type": "boolean",
+          "description": "If public IP is requested or not for the coordinator."
+        },
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1875]: recommend updating all these booleans to enum as explained here : 

[https://eng.ms/docs/products/arm/rp_onboarding/process/api_review_best_practices](https://eng.ms/docs/products/arm/rp_onboarding/process/api_review_best_practices)

Replace boolean/switch properties with better enum
A Boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Even if you still believe [True, False] are the correct values for a property, you should use a string enum with values [True, False] instead of boolean. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [True, False, Unknown].

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 9]
```diff
@@ -0,0 +1,2987 @@
...
+          "x-ms-mutability": [
+            "read",
+            "create"
+          ],
+          "description": "The geo-location where the resource lives"
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "../../../../../common-types/resource-management/v4/types.json#/definitions/Resource"
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1956]: >Resource [](http://example.com/codeflow?start=94&length=8)

Use ProxyResource or TrackedResource instead

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 10]
```diff
@@ -0,0 +1,2987 @@
...
+        },
+        "nextLink": {
+          "description": "Link to retrieve next page of results.",
+          "type": "string",
+          "readOnly": true
+        }
+      },
+      "description": "A list of private endpoint connections."
+    },
+    "PrivateLinkResource": {
```
Comment (raosuhas) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:2766]: >PrivateLinkResource [](http://example.com/codeflow?start=5&length=19)

use the definitions from here instead : 

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v4/privatelinks.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v4/privatelinks.json)

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 8 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1951]: Changed boolean types to enum

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 9 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json]: Used TrackedResource

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 10 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json]: Done

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 6 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:994]: In this case for 200 response code, since we only have "password" value in Role property, if we return 200 for no-op cases it will be causing security vulnerability.  I.e if endpoint returns 200 for the actual password, the password will be revealed.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 5 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:665]: Updated endpoint

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 4 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:504]: We have to keep the serverGroupsv2 on url because the costing is really high to change it currently. The PM team is aware and they want to proceed with the "cluster" naming instead of server group.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 7 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1072]: We have to keep the serverGroupsv2 on url because the costing is really high to change it currently. The PM team is aware of the change and they want to proceed with the "cluster" naming instead of server group.

**Comment by visingla-ms** [comment]
>         "operationId": "Configurations_UpdateNode",

It should be "Configurations_CreateorUpdateNode" or "Configurations_CreateNode".

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:670 in e455a41. [](commit_id = e455a41b6c519ea1f68b8f9d32e12e198a3626cf, deletion_comment = False)

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 6 above)
Comment (visingla-ms) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:994]: See the use of this: [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret).

**Comment by visingla-ms** [comment]
>           "202": {

You should 200 response as well.

Your previous comment here was: "In this case for 200 response code, since we only have "password" value in Role property, if we return 200 for no-op cases it will be causing security vulnerability.  I.e if endpoint returns 200 for the actual password, the password will be revealed."

See the use of this: [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret).

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:994 in e455a41. [](commit_id = e455a41b6c519ea1f68b8f9d32e12e198a3626cf, deletion_comment = False)

**Comment by visingla-ms** [comment]
>   "swagger": "2.0",

There are many Linter Issues, please fix these.

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:2 in e455a41. [](commit_id = e455a41b6c519ea1f68b8f9d32e12e198a3626cf, deletion_comment = False)

**Comment by senatol** [comment]
> > "swagger": "2.0",
> 
> There are many Linter Issues, please fix these.
> 
> Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:2 in [e455a41](https://github.com/Azure/azure-rest-api-specs/commit/e455a41b6c519ea1f68b8f9d32e12e198a3626cf). [](commit_id = [e455a41](https://github.com/Azure/azure-rest-api-specs/commit/e455a41b6c519ea1f68b8f9d32e12e198a3626cf), deletion_comment = False)

I've cleared most of the Lint warnings, for the warnings that are left:

AllResourcesMustHaveGetOperation - This warning for Role, We don't have a GET endpoint for Role on our service. GET operation is not functional for our case because it doesn't have any other property than password and the endpoint doesn't have any use case. The other warning for ServerConfiguration, we have another endpoint to get configurations.
PutRequestResponseScheme - We don't send adminPassword in the Cluster response which is the difference between request and response structures. For the other warnings we have non-settable properties on the schemas that are not used in request bodies but will exist on the responses.

Can you clarify if any of these are blocker for the PR? We are on a tight schedule and need to merge the PR as soon as possible.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json [hunk 11]
```diff
@@ -0,0 +1,2916 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Role"
+            }
+          },
+          "202": {
```
Comment (visingla-ms) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1003]: >202 [](http://example.com/codeflow?start=11&length=3)

PUT Request shouldn't have 202 response. Use 201, even for Async Response. See here: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creating-or-updating-resources-asynchronously](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creating-or-updating-resources-asynchronously).

**Comment by visingla-ms** [comment]
> 
> I've cleared most of the Lint warnings, for the warnings that are left:
> 
> AllResourcesMustHaveGetOperation - This warning for Role, We don't have a GET endpoint for Role on our service. GET operation is not functional for our case because it doesn't have any other property than password and the endpoint doesn't have any use case. The other warning for ServerConfiguration, we have another endpoint to get configurations. PutRequestResponseScheme - We don't send adminPassword in the Cluster response which is the difference between request and response structures. For the other warnings we have non-settable properties on the schemas that are not used in request bodies but will exist on the responses.
> 
> Can you clarify if any of these are blocker for the PR? We are on a tight schedule and need to merge the PR as soon as possible.

These are blocker issues.
1. PutRequestResponseSchema: You can tag the password property with x-ms-secret and omit it from Put Response Schema. Clients are not expected to receive it as it would be tagged as secret. So, you should be able to keep Put request and response schema same. ARM RPC requires PUT to have same request and response schema.
2. AllResourcesMustHaveGetOperation:
  a. Role resource: ARM requires all the top-level and nested resources to have GET operation. There are many ARM-ecosystem services need every ARM resource to have GET be present. Even if there is no valuable data to be returned today, could be in future. You can omit the secret properties in GET response, however, GET API is needed.
  b. ServerConfiguration resource: Same explanation as above. And I didn't understand that you have another endpoint to get configurations. Any Tracked/Proxy resource need to have at-least GET.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/ClusterCreate.json [hunk 12]
```diff
@@ -0,0 +1,166 @@
+{
+  "parameters": {
+    "subscriptionId": "ffffffff-ffff-ffff-ffff-ffffffffffff",
+    "resourceGroupName": "TestGroup",
+    "clusterName": "hsctestsg",
+    "api-version": "2022-11-08",
+    "parameters": {
+      "location": "westus",
+      "properties": {
+        "administratorLoginPassword": "p@ssword1234",
```
Comment (visingla-ms) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/ClusterCreate.json:10]: >p@ssword1234 [](http://example.com/codeflow?start=39&length=12)

CredScan Test is failing due to the sample passwords mentioned in examples. I would recommend to remove these sample passwords in example, and keep these properties empty in examples.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/RoleCreate.json [hunk 13]
```diff
@@ -0,0 +1,56 @@
+{
+  "parameters": {
+    "subscriptionId": "ffffffff-ffff-ffff-ffff-ffffffffffff",
+    "resourceGroupName": "TestGroup",
+    "clusterName": "pgtestsvc4",
+    "roleName": "role1",
+    "api-version": "2022-11-08",
+    "parameters": {
+      "properties": {
+        "password": "p@ssword1234"
```
Comment (visingla-ms) [inline specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/RoleCreate.json:10]: >p@ssword1234 [](http://example.com/codeflow?start=21&length=12)

CredScan Test is failing due to the sample passwords mentioned in examples. I would recommend to remove these sample passwords in example, and keep these properties empty in examples.

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 11 above)
Comment (mozansaka) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:1003]: @visingla-ms Hello, the way our Control Plane's been designed and been working for the last couple years is:
- For our main tracked resource: Cluster (previously called server group), PUT requests return 201 / 200 + ProvisioningState as the doc you referenced state as preferred model
- For other proxy resources Role and Firewall Rule, PUT requests return 202 Accepted. Users then can monitor the state of request via Azure-AsyncOperation, Location headers. We don't keep ProvisioningState at all and even though these are async requests, they are expected to finish fairly quickly. It's been like this for multiple years and one of our internal users "Azure Portal" team also has been using these endpoints.

Is this also blocker from arm review perspective or is it possible to add 201 as placeholder here and work on updating our backend model later?

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/ClusterCreate.json (hunk 12 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/ClusterCreate.json]: Fixed passwords

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/RoleCreate.json (hunk 13 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/examples/RoleCreate.json]: Fixed passwords

File: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json (hunk 11 above)
Comment (senatol) [reply specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json]: Updated PUT request responses to 201

**Comment by senatol** [comment]
@visingla-ms Hello, I've updated the PR with your feedbacks. 
We have only "PutRequestResponseScheme" warning for following resources.

For the **Cluster** resource, we have properties customers can't set, such as
- administratorLogin - customer receives the administrator login as default, they can't set this property,
- serverNames - it's provided after creation, 
- privateEndpointConnections - private endpoint connections are also not allowed to set  during creation.

For **Configuration,** the customer is allowed to set only "value" property, but configuration actually has other properties that customers are not allowed to change and we send them in the response:
- source
- description
- defaultValue
- dataType
- allowedValues
- requiresRestart

For **PrivateEndpointConnection**, we only have "privateLinkServiceConnectionState" property set by customer in request body and in the response we send:
- groupIds
- privateEndpoint 

Since request and response bodies are not coherent, we separated them to avoid the confusion for these properties which customers are not actually allowed to set.

**Comment by sjanamma** [comment]
> @visingla-ms Hello, I've updated the PR with your feedbacks. We have only "PutRequestResponseScheme" warning for following resources.
> 
> For the **Cluster** resource, we have properties customers can't set, such as
> 
> * administratorLogin - customer receives the administrator login as default, they can't set this property,
> * serverNames - it's provided after creation,
> * privateEndpointConnections - private endpoint connections are also not allowed to set  during creation.
> 
> For **Configuration,** the customer is allowed to set only "value" property, but configuration actually has other properties that customers are not allowed to change and we send them in the response:
> 
> * source
> * description
> * defaultValue
> * dataType
> * allowedValues
> * requiresRestart
> 
> For **PrivateEndpointConnection**, we only have "privateLinkServiceConnectionState" property set by customer in request body and in the response we send:
> 
> * groupIds
> * privateEndpoint
> 
> Since request and response bodies are not coherent, we separated them to avoid the confusion for these properties which customers are not actually allowed to set.

Our recommendation is to use the same definition for request / response of PUT operation. You should be able to re-use the model as long as readonly properties are marked properly.

**Comment by sjanamma** [comment]
>           "type": "string",

ProvisioningState must be enum. It should have at-least these 3 values: Succeeded, Failed, Canceled.

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:2248 in 7fd77d2. [](commit_id = 7fd77d2b8c1957a52e950eeb838fa742c0829163, deletion_comment = False)

**Comment by sjanamma** [comment]
>           "type": "boolean",

We suggest switching all booleans to enums to be able to extend them in future. ARM recommends enums over booleans for future proof APIs.

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/stable/2022-11-08/postgresqlhsc.json:2296 in 7fd77d2. [](commit_id = 7fd77d2b8c1957a52e950eeb838fa742c0829163, deletion_comment = False)

**Comment by senatol** [comment]
About switching booleans to enum, we have been using these booleans within our service private preview endpoint for almost 3 years now, Azure Portal being one of our internal users (+ some of the external customers). Also confirmed we don't have any plans to use any values other than [true, false] for them. Looking at our sister teams Cosmos DB [[Hub Generated] Review request for Microsoft.DocumentDB to add version stable/2022-08-15 by ashwinisingh01 · Pull Request #20405 · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/pull/20405) and Flexible Server PostgreSQL [Review request for Microsoft.DBforPostgreSQL Flexible server to add version stable/2022-12-01 by ambrahma · Pull Request #21803 · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/pull/21803), they both are still using booleans in their latest stable api versions and were able to get a sign off; so assuming this is not a blocker, we want to keep the booleans.

**Comment by live1206** [comment]
@tadelesh Could you help review the breaking change of go? Thanks!

**Comment by live1206** [comment]
@msyyc Could you help review the breaking change of python? Thanks!

**Review by live1206 — APPROVED** [review]
